migrate to tlv for invoice body

cfromknecht opened the pull request:
Builds on #3685 

This PR migrates our invoices to use TLV for their bodies. In the process we add feature vectors and payment addrs to make way for our mpp series. The unused receipt field is ultimately dropped from the db and the rpc calls. Finally, invoices are now populated with an empty set of features, which will not appear on payment requests until a non-zero number of bits are initially set.

joostjager commented on lnrpc/rpc.proto:
```diff
@@ -2515,6 +2515,13 @@ message PayReq {
+message Feature {
+        uint32 bit = 1 [json_name = "bit"];
+        string name = 2 [json_name = "name"];
+        bool is_required = 3 [json_name = "is_required"];
```
Looks like feature bits can be its own pr

cfromknecht replied:
i kept the changes to populate an empty feature vector, but will move the rest of it to another pr

halseth commented on channeldb/migration12/invoices.go:
```diff
@@ -0,0 +1,475 @@
+// Note: this function is in use for a migration. Before making changes that
+// would modify the on disk format, make a copy of the original code and store
+// it with the migration.
+func serializeInvoice(w io.Writer, i *Invoice) error {
```
This code does look new. It should be added to the next commit to make it clear what code has just been copied and what is new (and hence needs a more close review)

halseth commented on channeldb/migration12/migration.go:
```diff
@@ -0,0 +1,74 @@
+// MigrateInvoiceTLV migrates all existing invoice bodies over to be serialized
+// in a single TLV stream. In the process, we drop the Receipt field and add
+// PaymentAddr and Features to the invoice Terms.
+func MigrateInvoiceTLV(tx *bbolt.Tx) error {
```
this should have a test.

cfromknecht replied:
done

cfromknecht commented on cmd/lncli/commands.go:
```diff
@@ -2561,10 +2561,6 @@ var addInvoiceCommand = cli.Command{
-		cli.StringFlag{
-			Name:  "receipt",
-			Usage: "an optional cryptographic receipt of payment",
-		},
```
done

cfromknecht commented on rpcserver.go:
```diff
@@ -3525,7 +3525,6 @@ func (r *rpcServer) AddInvoice(ctx context.Context,
 
 	addInvoiceData := &invoicesrpc.AddInvoiceData{
 		Memo:            invoice.Memo,
-		Receipt:         invoice.Receipt,
```
done

cfromknecht commented on channeldb/invoices.go:
```diff
@@ -154,8 +164,19 @@ type ContractTerm struct {
+	PaymentAddr [32]byte
+
+	// Features is the feature vectors advertised on the payment request.
+	Features *lnwire.FeatureVector
```
field restructuring is now is now in a separate commit

cfromknecht commented on channeldb/invoices.go:
```diff
@@ -154,8 +164,19 @@ type ContractTerm struct {
+
+	// PaymentAddr is a randomly generated value include in the MPP record
+	// by the sender to prevent probing of the receiver.
+	PaymentAddr [32]byte
```
i'd rather add these fields during the migration so that users can't accidentally revert to a commit where we don't process them. the downgrade logic in opening channeldb will prevent this if they are added here

joostjager replied:
Does this mean that we are going to increase the db version with every tlv field that we add, a no-op migration>

cfromknecht replied:
i think it's the safest path

joostjager replied:
Ok. It is easy to forget though. Should we add a NOTE comment to every list of db tlv type constants in the code base to remind future devs about this? Pre-existing, but maybe not worth the overhead of a separate pr

cfromknecht replied:
done

cfromknecht commented on channeldb/migration12/invoices.go:
```diff
@@ -0,0 +1,475 @@
+		paymentAddrType tlv.Type = 21
+		featuresType    tlv.Type = 23
+		stateType       tlv.Type = 25
+		amtPaidType     tlv.Type = 27
```
it's purely cosmetic, tho likely that #3701 will be merged before at which point we can reevaluate the constants

joostjager replied:
It has been merged, so the option is available to do cosmetics

cfromknecht replied:
done

cfromknecht commented on channeldb/migration12/invoices.go:
```diff
@@ -0,0 +1,475 @@
+	acceptTimeType   tlv.Type = 9
+	resolveTimeType  tlv.Type = 11
+	expiryHeightType tlv.Type = 13
+	stateType        tlv.Type = 15
```
given that we do have additional data, i don't see how we could intermingle them without running into issues. 

> Maybe passing through every custom tlv fields in the 2^16+ range works and then doing a custom mapping for the spec fields that we know of?

are you referring to the tlvs presented in the onion payload? for now those are kept in the onion, but if we wanted to persist them separately i imagine adding another field for all the experimental types

joostjager replied:
so you mean a single tlv field that contains another tlv array?

joostjager replied:
Yes, indeed referring to the custom fields in the onion payload

cfromknecht replied:
> so you mean a single tlv field that contains another tlv array?

possibly yes, or appending it after all the htlcs

joostjager replied:
ok, something to think about later.

cfromknecht commented on .golangci.yml:
```diff
@@ -12,6 +12,7 @@ run:
 
   skip-dirs:
     - channeldb/migration_01_to_11
+    - channeldb/migration12
```
ended up adding this to the list of linted packages, fixed any other linter errors and added the `nolint: dupl` directives for the methods in this package

cfromknecht commented on lnwire/features.go:
```diff
@@ -349,10 +348,10 @@ func (fv *FeatureVector) UnknownRequiredFeatures() []FeatureBit {
+		return name
 	}
-	return fmt.Sprintf("%s(%d)", name, bit)
+	return "unknown"
```
will update in separate pr

cfromknecht commented on lnrpc/invoicesrpc/invoices_server.go:
```diff
@@ -247,13 +247,14 @@ func (s *Server) AddHoldInvoice(ctx context.Context,
+		MaxPaymentMSat:     s.cfg.MaxPaymentMSat,
+		DefaultCLTVExpiry:  s.cfg.DefaultCLTVExpiry,
+		ChanDB:             s.cfg.ChanDB,
+		GenInvoiceFeatures: s.cfg.GenInvoiceFeatures,
```
they are constant, but i went with the approach of generating a new vector for each invoice to avoid any possibility of mutations

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -105,12 +112,6 @@ type Invoice struct {
-	//
-	// TODO(roasbeef): document scheme.
-	Receipt []byte
-
```
Wasn't `Receipt` already gone?

cfromknecht replied:
not from the invoice, it was still present on the data structure itself since the migration hadn't been performed. prior it was removed from the rpc and everywhere outside of channeldb

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -345,71 +346,90 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
-	}
-
-	if err := binary.Write(w, byteOrder, int64(i.Expiry)); err != nil {
-		return err
```
Maybe this shouldn't have been copied in the previous commit?

cfromknecht replied:
done

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -345,71 +346,90 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
+
+		// Invoice state.
+		tlv.MakePrimitiveRecord(stateType, &state),
+		tlv.MakePrimitiveRecord(amtPaidType, &amtPaid),
```
Difficult to see whether this is exactly what will be added in a later commit in `channeldb`

joostjager replied:
But probably nothing we can do about it. Hopefully this is the last invoice migration

joostjager commented on channeldb/invoices.go:
```diff
@@ -174,12 +191,6 @@ type Invoice struct {
-	//
-	// TODO(roasbeef): document scheme.
-	Receipt []byte
-
```
Also not removed in first commit?

cfromknecht replied:
we could remove it sooner and instead write a blank receipt, but it is done here because here this is the commit that actually modifies the on-disk format. if we do it sooner then the existing serialization code can't reference any (in theory) existing receipts, and means that there would side effects if an invoice is decoded and deserialized before running the migration.

joostjager replied:
ok

joostjager commented on channeldb/invoices.go:
```diff
@@ -312,16 +323,14 @@ func validateInvoice(i *Invoice) error {
 	}
+	if i.Terms.Features == nil {
+		return errors.New("invoice must have a feature vector")
+	}
```
Can the addition of features be separated from the db changes?

cfromknecht replied:
in what sense? another pr?

joostjager replied:
i thought separate commit. isolate addition of the feature from moving to tlv. but nvm, it doesn't really matter

joostjager commented on channeldb/migration12/migration_test.go:
```diff
@@ -0,0 +1,306 @@
+		0xbc, 0xbf, 0x72, 0x4e, 0x1e, 0x00, 0x0b, 0x08, 0x00, 0x17,
+		0xaf, 0x4c, 0x22, 0xc4, 0x24, 0x00, 0x0d, 0x04, 0x00, 0x00,
+		0x23, 0x1d, 0x0f, 0x01, 0x02,
+	}
```
This is indeed an idea too, to prevent even more serialize code.

cfromknecht replied:
yes this prevented me from having to copy in the old serialization code :)

joostjager commented on channeldb/migtest/migtest.go:
```diff
@@ -0,0 +1,94 @@
+
+// ApplyMigration is a helper test function that encapsulates the general steps
+// which are needed to properly check the result of applying migration function.
+func ApplyMigration(t *testing.T,
```
Ref this in `migration_01_to_11`? Maybe not...

cfromknecht replied:
in that package the after and before functions operate on `channeldb.DB`, where as here they operate on `bbolt.Tx`. some of those functions rely on db-level methods, so it is a bigger refactor. i made the decision here to only offer `bbolt.Tx` because it makes the closures more isolated, and hopefully less code to copy in future ones.

joostjager replied:
sounds good

joostjager commented on channeldb/invoices.go:
```diff
@@ -1121,7 +1121,7 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
 	for {
 		// Read the length of the tlv stream for this htlc.
-		var streamLen uint64
+		var streamLen int64
```
typo in commit msg

joostjager commented:
Build generates a diff in `go.mod`

cfromknecht commented:
> This is the ListInvoices diff.

Looked into this, the `channeldb.Invoice` struct was still using `invoice.Terms.State` for encode decode. Fixed in latest commit by pointing it to `i.State` and removing the old field.

> Build generates a diff in go.mod

also fixed, a `log.Error` in `migtest` caused it to import the prometheus log. it has not been replaced with `t.Fatalf`

cfromknecht commented on subrpcserver_config.go:
```diff
@@ -210,6 +212,9 @@ func (s *subRPCServerConfigs) PopulateDependencies(cc *chainControl,
 			)
+			subCfgValue.FieldByName("GenInvoiceFeatures").Set(
+				reflect.ValueOf(genInvoiceFeatures),
+			)
```
yes that sounds like a good idea, since we need it for the main rpc

joostjager commented:
```
--- FAIL: TestSettleInvoice (0.01s)
    invoiceregistry_test.go:94: invoice must have a feature vector
```

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -36,7 +36,22 @@ const (
+	paymentAddrType tlv.Type = 10
+	featuresType    tlv.Type = 11
+	invStateType    tlv.Type = 12
+	amtPaidType     tlv.Type = 13
```
Beautiful

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -36,7 +36,22 @@ const (
 	resolveTimeType  tlv.Type = 11
 	expiryHeightType tlv.Type = 13
-	stateType        tlv.Type = 15
+	htlcStateType    tlv.Type = 15
```
I don't think we want to use this migration where we write out all htlcs again to beautify these constant too?

cfromknecht replied:
we definitely can, but it adds nothing functionally. i'd rather save this for the next migration where we do for htlcs since it is purely cosmetic and at this point can only introduce more bugs

joostjager replied:
agrred

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -339,7 +361,73 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
+		return err
+	}
+
+	b.Reset()
```
Hmm, isn't this going to overwrite the underlying array of `featureBytes`? If not, maybe add a comment stating this

cfromknecht replied:
TIL! yes it does, tho fortunately the empty features serialize to zero bytes and so don't affect the migration. however, this was copied into the main serialization method, so that could have been an issue but i've modified both to use a distinct buffer.

cfromknecht replied:
https://play.golang.org/p/kcNLH-34qH4

joostjager commented on channeldb/invoices.go:
```diff
@@ -96,14 +91,40 @@ const (
+	// NOTE: A migration should be added whenever this list changes. This
+	// prevents against the database being rolled back to an older
+	// format where the surrounding logic might assume a different set of
+	// fields are known.
```
:+1:
It is a no-op migration, hopefully that is clear.

cfromknecht replied:
it is only a noop migration if the field is truly optional, in most cases that won't be the case i presume

joostjager replied:
hmm, so in order to minimize migrations, what would we do with for example a new required int field? write a migration or define non-present as 0? i was hoping we could have fewer migrations now, but maybe not

cfromknecht replied:
if it is required, idt there's a way around it. some things may be less critical and it would be okay to do an in-flight migration. in this case tho, these parameters affect invoice settling and i think qualify as required. i think it will have to be a case by case basis

joostjager commented on channeldb/invoice_test.go:
```diff
@@ -114,7 +114,7 @@ func TestInvoiceWorkflow(t *testing.T) {
 		t.Fatalf("unable to fetch invoice: %v", err)
 	}
-	if dbInvoice2.Terms.State != ContractSettled {
+	if dbInvoice2.State != ContractSettled {
```
Commit message describes move of final delta and expiry, but that doesn't happen in this commit anymore

cfromknecht replied:
good catch, restored

halseth commented on channeldb/migration12/invoices.go:
```diff
@@ -0,0 +1,392 @@
+//
+// nolint: dupl
+func serializeInvoice(w io.Writer, i *Invoice) error {
+	panic("unimplemented")
```
I think the method can just be removed, it is not used ion this commit.

cfromknecht replied:
none of the code in this commit is used

cfromknecht replied:
refactored and now `SerializeInvoice` is always used

halseth commented on channeldb/invoices.go:
```diff
@@ -1141,11 +1141,9 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
-		streamReader := bytes.NewReader(streamBytes)
+		// Limit the reader so that it stops at the end of this htlc's
+		// stream.
+		htlcReader := io.LimitReader(r, streamLen)
```
is this in any hot-path?

cfromknecht replied:
not that i know of, but it will be as invoices start to have multiple htlcs or payments. might improve `listinvoices` performance slightly

halseth commented on channeldb/invoices.go:
```diff
@@ -166,6 +166,13 @@ func (c ContractState) String() string {
 type ContractTerm struct {
+	// FinalCltvDelta is the minimum required number of blocks before htlc
+	// expiry when the invoice is accepted.
+	FinalCltvDelta int32
```
Should we use the opportunity to remove the `ContractTerm` struct? Doesn't seem it add much, it only adds to the confusion where to put the fields. 

As it is not serialized separately either, I would prefer having all these fields on the invoice instead.

Roasbeef replied:
I'd be down with removing it, it was created back in a time when the realm of possibilities was a bit more unbounded...🌈

cfromknecht replied:
imo like having a separation of invoice parameters and invoice metadata. the latest restructuring moves the immutable constraints into `ContractTerm`, which are ultimately embedded in the payment request. things like `State`, which is a) mutable, and b) not present in the payreq, is now contained on the main body with the other metadata.

if we don't think this adds any value tho we can flatten entirely.

cfromknecht replied:
ultimately tho this doesn't affect the migration at all, since we have a working copy of the serialization in `channeldb` and are free to change the in-memory structure until the next one

halseth replied:
If we are to keep the `ContractTerms`, I think there are a few more immutable things to move there? pay req, memo perhaps. I would still prefer just flatten it, but not a blocker  :)

halseth commented on channeldb/migration12/invoices.go:
```diff
@@ -0,0 +1,572 @@
+)
+
+const (
+	// MaxMemoSize is maximum size of the memo field within invoices stored
```
what have we gotten ourself into 😅

Roasbeef replied:
?

halseth commented on channeldb/invoices.go:
```diff
@@ -892,71 +912,73 @@ func putInvoice(invoices, invoiceIndex, addIndex *bbolt.Bucket,
-	}
-
-	birthBytes, err := i.CreationDate.MarshalBinary()
+	creationDateBytes, err := i.CreationDate.MarshalBinary()
```
didn't we want to import this method from the v12 migration to avoid the duplication?

cfromknecht replied:
done

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -36,7 +36,22 @@ const (
+	paymentAddrType tlv.Type = 10
+	featuresType    tlv.Type = 11
+	invStateType    tlv.Type = 12
+	amtPaidType     tlv.Type = 13
```
Can we use this migration to also add the payment hash as a proper field here? Especially for hodl invoices, there is some awkwardness to get to that value.

joostjager replied:
Idea came up in conversation with @bhandras

cfromknecht replied:
the payment hash is already serialized in the invoice index, so preferably we would just extract it from there instead of serializing again. unfortunately we only track `hash => invoice number`, so it's not so simple to recover the hash from the invoice on it's own.

> Especially for hodl invoices, there is some awkwardness to get to that value.

How is it extracted today?

joostjager replied:
In `CreateRPCInvoice`, the hash is decoded again from the payment request. But maybe we should only add `PaymentHash` to the in-memory invoice and make sure we always populate it in `LookupInvoice`, `FetchAllInvocies`, `QueryInvoices`, etc?

cfromknecht replied:
yeah i think that'd better than serializing it again on the invoice, though it may be simpler to do that

cfromknecht replied:
i think that can be done without a migration, so it should be okay to defer for now

Roasbeef commented on channeldb/invoices.go:
```diff
@@ -96,14 +91,40 @@ const (
+
+	// A set of tlv type definitions used to serialize invoice bodiees.
+	//
+	// NOTE: A migration should be added whenever this list changes. This
```
The migration would only just update the DB version to prevent a downgrade? Seems like we could also implement a run-time sanity check based on if the DB includes a type not in the set of known types, or the other way around.

cfromknecht replied:
> The migration would only just update the DB version to prevent a downgrade? 

There will also be code added to read and write the new field

> Seems like we could also implement a run-time sanity check based on if the DB includes a type not in the set of known types, or the other way around.

Does that mean we'd read the entire db on every open? or only do it when the data is accessed? i think it ultimately depends on what the field is, in certain cases we could lose money if a downstream action fails that we are expecting to succeed. the migration is def the safest option because it doesn't create the illusion that things are okay

Roasbeef commented on channeldb/invoices.go:
```diff
@@ -174,23 +206,10 @@ type Invoice struct {
 	// or any other message which fits within the size constraints.
 	Memo []byte
 
-	// Receipt is an optional field dedicated for storing a
```
R.I.P 😭

Roasbeef commented on channeldb/invoices.go:
```diff
@@ -174,23 +206,10 @@ type Invoice struct {
-	FinalCltvDelta int32
-
-	// Expiry defines how long after creation this invoice should expire.
-	Expiry time.Duration
```
Don't we want to maintain these fields, as otherwise, the invoice needs to be decoded just to access them.

cfromknecht replied:
we still have the fields, they've just been moved

Roasbeef commented on channeldb/invoices.go:
```diff
@@ -223,6 +242,9 @@ type Invoice struct {
 	SettleIndex uint64
 
+	// State describes the state the invoice is in.
+	State ContractState
```
If the `ContractTerm` field is staying, then why move this _outside_ of it?

cfromknecht replied:
see the other inline comment re: `ContractTerm`, but the main rationale is that `State` is more metadata than it is a constraint on the sender.

Roasbeef commented on channeldb/invoices.go:
```diff
@@ -1018,79 +1040,89 @@ func fetchInvoice(invoiceNum []byte, invoices *bbolt.Bucket) (Invoice, error) {
 
-	if err := binary.Read(r, byteOrder, &invoice.FinalCltvDelta); err != nil {
-		return invoice, err
+	lr := io.LimitReader(r, bodyLen)
```
Nice usage of `LimitReader` 👍

Roasbeef commented on channeldb/migration12/invoices.go:
```diff
@@ -0,0 +1,572 @@
+// as a map.
+//
+// nolint: dupl
+func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
```
Since this didn't change, can't we use the version from `migration_1_to_11`?

joostjager replied:
or just don't deserialize/serialize the htlc list at all. just pass through the bytes

cfromknecht replied:
it does change w/ the `LimitReader` optimization

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -339,7 +361,73 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
+		return err
+	}
+
+	return serializeHtlcs(w, i.Htlcs)
```
Why don't we actually make this a tlv field?

joostjager replied:
Then it also wouldn't be necessary to start the serialized htlcs blob with a length.

cfromknecht replied:
> Then it also wouldn't be necessary to start the serialized htlcs blob with a length.

i think each one would still need a length, no? perhaps you mean the body wouldn't need to be length prefixed?

imo it'd be preferable to serialize each htlc under it's own circuit key instead, then we eliminate length-prefixed tlv streams altogether. do we think that's worth doing now?

joostjager replied:
Hmm, there are indeed multiple variations. But to add least add the htlc tlv blob as is as a tlv record on the invoice level looks more consistent to me.

cfromknecht replied:
i think for now we should leave them where they are, since for the time being that would _introduce_ a performance penalty in needed to allocate a backing slice for all htlcs. i'd rather not move them preemptively until we know what the workload/bottlenecks are surrounding it.

joostjager commented on channeldb/migration12/invoices.go:
```diff
@@ -339,7 +361,73 @@ func deserializeHtlcs(r io.Reader) (map[CircuitKey]*InvoiceHTLC, error) {
+		tlv.MakePrimitiveRecord(settleIndexType, &i.SettleIndex),
+
+		// Terms.
+		tlv.MakePrimitiveRecord(preimageType, &preimage),
```
I start to realize that this migration offers more opportunities to clean up some things. The situation with "preimage unknown" could be dealt with much more elegant by just not writing out the preimage tlv record.

For the add/settle metadata similar a similar thing could be possible.

I know we want to move on with the rest of the stack of mpp prs, but the effort required to fix these things now seems to be much less than when it would need to be a separate migration later on. What do you think?

cfromknecht replied:
i would vote to revisit these things if they become an issue, since those optimizations don't provide any tangible benefits in the grand scheme of things

cfromknecht commented on channeldb/channel.go:
```diff
@@ -101,8 +102,7 @@ var (
 	// decoded because the byte slice is of an invalid length.
-	ErrInvalidCircuitKeyLen = fmt.Errorf(
-		"length of serialized circuit key must be 16 bytes")
+	ErrInvalidCircuitKeyLen = migration12.ErrInvalidCircuitKeyLen
```
reverted to prior version of the pr

cfromknecht commented:
@joostjager @Roasbeef @halseth reverted to prior pr and touched up for another review. ended up using joost's suggestion of passing through the htlc bytes directly, which shaved 200+ lines off the diff

halseth commented on channeldb/migration12/invoices.go:
```diff
@@ -0,0 +1,240 @@
+// LegacyDeserializeInvoice decodes an invoice from the passed io.Reader using
+// the pre-TLV serialization.
+//
+// nolint: dupl
```
nolint not needed anymore?

halseth commented on channeldb/migration12/invoices.go:
```diff
@@ -225,6 +232,95 @@ func LegacyDeserializeInvoice(r io.Reader) (Invoice, error) {
+// DeserializeInvoice deserializes an invoice from an io.Reader.
+//
+// nolint: dupl
+func DeserializeInvoice(r io.Reader) (Invoice, error) {
```
is it needed for anything other than tests? Maybe we could just check the final bytes in the tests and just remove this method.

cfromknecht replied:
done
